File: agentprimordia/internal/chaos/report.go
package chaos

import (
	"fmt"
	"strings"
	"time"
)

// ===== 实验报告生成器 =====

// FormatReport 将实验结果格式化为 Markdown 报告
func FormatReport(result *ExperimentResult) string {
	var sb strings.Builder

	sb.WriteString("# 混沌实验报告\n\n")
	sb.WriteString(fmt.Sprintf("**实验名称**: %s\n\n", result.Experiment.Name))

	if result.Experiment.Description != "" {
		sb.WriteString(fmt.Sprintf("**描述**: %s\n\n", result.Experiment.Description))
	}

	sb.WriteString(fmt.Sprintf("**假设**: %s\n\n", result.Experiment.Hypothesis))

	statusEmoji := "✅"
	if !result.HypothesisValidated {
		statusEmoji = "❌"
	}

	sb.WriteString(fmt.Sprintf("**假设验证**: %s %s\n\n",
		statusEmoji,
		map[bool]string{true: "已验证", false: "未验证"}[result.HypothesisValidated],
	))

	sb.WriteString(fmt.Sprintf("**状态**: %s\n\n", result.Status))
	sb.WriteString(fmt.Sprintf("**持续时间**: %v\n\n", result.Duration))
	sb.WriteString(fmt.Sprintf("**开始时间**: %s\n\n", result.StartTime.Format(time.RFC3339)))
	sb.WriteString(fmt.Sprintf("**结束时间**: %s\n\n", result.EndTime.Format(time.RFC3339)))

	// 故障注入结果
	if len(result.FaultResults) > 0 {
		sb.WriteString("## 注入的故障\n\n")
		sb.WriteString("| # | 类型 | 描述 | 注入状态 | 注入时间 | 清理时间 | 错误 |\n")
		sb.WriteString("|---|------|------|----------|----------|----------|------|\n")
		for i, fr := range result.FaultResults {
			injected := "✅ 成功"
			if !fr.Injected {
				injected = "❌ 失败"
			}
			errMsg := ""
			if fr.Error != nil {
				errMsg = fr.Error.Error()
			}
			sb.WriteString(fmt.Sprintf("| %d | %s | %s | %s | %s | %s | %s |\n",
				i+1,
				fr.FaultType,
				fr.Description,
				injected,
				fr.InjectTime.Format("15:04:05"),
				func() string {
					if fr.CleanupTime.IsZero() {
						return "-"
					}
					return fr.CleanupTime.Format("15:04:05")
				}(),
				errMsg,
			))
		}
		sb.WriteString("\n")
	}

	// 稳态检查
	if result.Experiment.SteadyState != nil {
		sb.WriteString("## 稳态检查\n\n")
		sb.WriteString(fmt.Sprintf("**稳态条件**: %s\n\n", result.Experiment.SteadyState.Name()))

		sb.WriteString("### 实验前\n\n")
		formatSteadyStateResult(&sb, &result.PreSteadyState)

		sb.WriteString("### 实验后\n\n")
		formatSteadyStateResult(&sb, &result.PostSteadyState)
	}

	// 标签
	if len(result.Experiment.Tags) > 0 {
		sb.WriteString(fmt.Sprintf("\n**标签**: %s\n", strings.Join(result.Experiment.Tags, ", ")))
	}

	return sb.String()
}

func formatSteadyStateResult(sb *strings.Builder, r *SteadyStateResult) {
	metEmoji := "✅"
	if !r.Met {
		metEmoji = "❌"
	}
	sb.WriteString(fmt.Sprintf("- **状态**: %s %s\n", metEmoji,
		map[bool]string{true: "满足", false: "不满足"}[r.Met]))
	sb.WriteString(fmt.Sprintf("- **消息**: %s\n", r.Message))
	if len(r.Details) > 0 {
		sb.WriteString("- **详情**:\n")
		for k, v := range r.Details {
			sb.WriteString(fmt.Sprintf("  - %s: %v\n", k, v))
		}
	}
	sb.WriteString("\n")
}

// ExperimentSummary 实验摘要（用于批量报告）
type ExperimentSummary struct {
	Name                string
	Status              ExperimentStatus
	HypothesisValidated bool
	Duration            time.Duration
	FaultCount          int
	SteadyStateMet      bool
}

// Summarize 生成实验摘要
func Summarize(result *ExperimentResult) ExperimentSummary {
	return ExperimentSummary{
		Name:                result.Experiment.Name,
		Status:              result.Status,
		HypothesisValidated: result.HypothesisValidated,
		Duration:            result.Duration,
		FaultCount:          len(result.FaultResults),
		SteadyStateMet:      result.PostSteadyState.Met,
	}
}

// FormatSummaryTable 将多个实验摘要格式化为表格
func FormatSummaryTable(summaries []ExperimentSummary) string {
	var sb strings.Builder
	sb.WriteString("| 实验 | 状态 | 假设 | 持续时间 | 故障数 | 稳态 |\n")
	sb.WriteString("|------|------|------|----------|--------|------|\n")
	for _, s := range summaries {
		validated := "✅"
		if !s.HypothesisValidated {
			validated = "❌"
		}
		met := "✅"
		if !s.SteadyStateMet {
			met = "❌"
		}
		sb.WriteString(fmt.Sprintf("| %s | %s | %s | %v | %d | %s |\n",
			s.Name, s.Status, validated, s.Duration, s.FaultCount, met))
	}
	return sb.String()
}

File: sdk/typescript/src/chaos/report.ts
/**
 * report.ts — 实验报告生成
 *
 * 对齐 Go 端 internal/chaos/report.go
 * Stability: Experimental
 */

import type { ExperimentResult, ExperimentStatus } from './engine.js';

/** 实验摘要（用于批量报告，对齐 Go ExperimentSummary） */
export interface ExperimentSummary {
  name: string;
  status: ExperimentStatus;
  hypothesisValidated: boolean;
  durationMs: number;
  faultCount: number;
  steadyStateMet: boolean;
}

/** 生成实验摘要（对齐 Go Summarize） */
export function summarize(result: ExperimentResult): ExperimentSummary {
  return {
    name: result.experiment.name,
    status: result.status,
    hypothesisValidated: result.hypothesisValidated,
    durationMs: result.durationMs,
    faultCount: result.faultResults.length,
    steadyStateMet: result.postSteadyState?.met ?? true,
  };
}

/** 将实验结果格式化为 Markdown 报告（对齐 Go FormatReport） */
export function formatReport(result: ExperimentResult): string {
  const lines: string[] = [];

  lines.push('# 混沌实验报告');
  lines.push('');
  lines.push(`**实验名称**: ${result.experiment.name}`);
  lines.push('');

  if (result.experiment.description) {
    lines.push(`**描述**: ${result.experiment.description}`);
    lines.push('');
  }

  lines.push(`**假设**: ${result.experiment.hypothesis}`);
  lines.push('');

  const statusIcon = result.hypothesisValidated ? '✅' : '❌';
  lines.push(`**假设验证**: ${statusIcon} ${result.hypothesisValidated ? '已验证' : '未验证'}`);
  lines.push('');
  lines.push(`**状态**: ${result.status}`);
  lines.push('');
  lines.push(`**持续时间**: ${result.durationMs}ms`);
  lines.push('');
  lines.push(`**开始时间**: ${result.startTime.toISOString()}`);
  lines.push('');
  lines.push(`**结束时间**: ${result.endTime.toISOString()}`);
  lines.push('');

  // 故障注入结果
  if (result.faultResults.length > 0) {
    lines.push('## 注入的故障');
    lines.push('');
    lines.push('| # | 类型 | 描述 | 注入状态 | 注入时间 | 清理时间 | 错误 |');
    lines.push('|---|------|------|----------|----------|----------|------|');
    result.faultResults.forEach((fr, i) => {
      const injected = fr.injected ? '✅ 成功' : '❌ 失败';
      const injectTime = fr.injectTime.toLocaleTimeString();
      const cleanupTime = fr.cleanupTime ? fr.cleanupTime.toLocaleTimeString() : '-';
      const errMsg = fr.error?.message ?? '';
      lines.push(`| ${i + 1} | ${fr.faultType} | ${fr.description} | ${injected} | ${injectTime} | ${cleanupTime} | ${errMsg} |`);
    });
    lines.push('');
  }

  // 稳态检查
  if (result.experiment.steadyState) {
    lines.push('## 稳态检查');
    lines.push('');
    lines.push(`**稳态条件**: ${result.experiment.steadyState.name()}`);
    lines.push('');

    lines.push('### 实验前');
    lines.push('');
    if (result.preSteadyState) {
      formatSteadyStateResult(lines, result.preSteadyState);
    }

    lines.push('### 实验后');
    lines.push('');
    if (result.postSteadyState) {
      formatSteadyStateResult(lines, result.postSteadyState);
    }
  }

  // 标签
  if (result.experiment.tags && result.experiment.tags.length > 0) {
    lines.push('');
    lines.push(`**标签**: ${result.experiment.tags.join(', ')}`);
  }

  return lines.join('\n');
}

function formatSteadyStateResult(lines: string[], r: { met: boolean; message: string; details?: Record<string, unknown> }): void {
  const metIcon = r.met ? '✅' : '❌';
  lines.push(`- **状态**: ${metIcon} ${r.met ? '满足' : '不满足'}`);
  lines.push(`- **消息**: ${r.message}`);
  if (r.details && Object.keys(r.details).length > 0) {
    lines.push('- **详情**:');
    for (const [k, v] of Object.entries(r.details)) {
      lines.push(`  - ${k}: ${v}`);
    }
  }
  lines.push('');
}

/** 将多个实验摘要格式化为表格（对齐 Go FormatSummaryTable） */
export function formatSummaryTable(summaries: ExperimentSummary[]): string {
  const lines: string[] = [];
  lines.push('| 实验 | 状态 | 假设 | 持续时间 | 故障数 | 稳态 |');
  lines.push('|------|------|------|----------|--------|------|');

  for (const s of summaries) {
    const validated = s.hypothesisValidated ? '✅' : '❌';
    const met = s.steadyStateMet ? '✅' : '❌';
    lines.push(`| ${s.name} | ${s.status} | ${validated} | ${s.durationMs}ms | ${s.faultCount} | ${met} |`);
  }

  return lines.join('\n');
}
